rust/guards/src/lib.rs

```diff
@@ -0,0 +1,164 @@
+    /// Items that are currently being processed.
+    processing_items: BTreeSet<String>,
+}
+
+thread_local! {
+    static STATE: RefCell<State> = RefCell::new(State::default());
+}
+
+#[update]
+async fn process_single_item_with_panicking_callback(
+    item_to_process: String,
+    future_type: FutureType,
+) {
+    let _prevent_parallel_calls_guard = ensure_not_in_processing(item_to_process.clone());
+    ensure_not_processed(&item_to_process);
+    let _process_item_only_once_guard = scopeguard::guard((), |_| {
+        STATE.with(|state| {
+            state
+                .borrow_mut()
+                .items
+                .entry(item_to_process)
+                .and_modify(|v| *v = true);
+        });
+    });
+
+    // Call and await future. There are 2 scenarios to consider.
+    // 1) The future cannot be polled until completion.
+    // The future returns a result of type `futures::task::Poll::Pending` indicating that it's not ready yet.
+    // In that case, the state will be committed and execution will yield, terminating the execution of that first message.
+    // Execution will then continue in a second message, when the future is ready.
+    // That means that the panic at the end will only revert the state changes occurring in the second message.
+    // The Rust ic_cdk will during `call_on_cleanup` call `Drop` on any variables that still in scope at the end of the first message,
+    // hence the guard will be executed and not reverted.
+    //
+    // 2) The future can be polled until completion.
+    // The future returns a result of type `futures::task::Poll::Ready` and in that case execution will continue without yielding.
+    // Everything will be executed in a single message and any state modification will be dropped due to the panic occurring at the end.
+    // In that case, the guard is ineffective
+    future_type.call().await;
+
+    panic!("panicking callback!")
+}
+
+#[update]
+async fn process_all_items_with_panicking_callback(
+    panicking_item: String,
+    future_type: FutureType,
+) {
+    let items: Vec<String> = STATE.with(|state| state.borrow().items.keys().cloned().collect());
+    for item in items {
+        let _prevent_parallel_calls_guard = ensure_not_in_processing(item.clone());
+        ensure_not_processed(&item);
+        let _process_item_only_once_guard = scopeguard::guard((), |_| {
+            STATE.with(|state| {
+                state
+                    .borrow_mut()
+                    .items
+                    .entry(item.clone())
+                    .and_modify(|v| *v = true);
+            });
+        });
+        // Note that this is the callback of the previous item. Similarly to the above method,
+        // the same 2 scenarios apply and whether the guard is effective or not depends on whether
+        // the future was polled until completion in a single message.
+        if item == panicking_item {
+            panic!("panicking callback!");
+        }
```

Comment:
I'm wondering if this can be a bit confusing for readers since it happens before the `await` point. I suppose if the `panicking_item` is the first in the list, then the code is not very meaningful -- anything else besides the first item is meaningful.

Reply:
> I'm wondering if this can be a bit confusing for readers since it happens before the await point.

Initially this originated from a conversation I had with Thomas M. asking whether panicking at step i would impact the state of the elements processed during previous steps i-1 ...1,0. This is what the comment above L.72 tries to explain `// Note that this is the callback of the previous item.` Do you think this needs more explanation?.

> I suppose if the panicking_item is the first in the list, then the code is not very meaningful

Absolutely. We could add a check that this is not the case, but this might "muddy" the message in the sense that this would be an additional guard that's not really useful to carry the point of this example. So I would tend to leave as is, WDYT?

Reply:
We could enhance the comment to mention explicitly that if the panicking item is the first in the list, then it becomes kind of a no-op but any other element creates the same situation as the above method.

This is not a strong opinion though, so we could also leave things as they are.